### PR TITLE
Parse sam metadata from sam files

### PR DESCRIPTION
**Pull request by bernt-matthias**
Many tools support Sam and bam input. If the tool uses metadata, e.g. for dynamic options, it's useful that both datatypes support the same set of metadata. 

Code is copied from  BamNative. 

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Review by mvdbeek — COMMENTED** [review]
I think the symlinking there didn't work:
```
>               raise AssertionError(f"Test input file ({filename}) cannot be found.")
E               AssertionError: Test input file (sam_with_header.sam) cannot be found.
```

**Comment by bernt-matthias** [comment]
Lets see if it works the other way round. 

One thing that is not yet optimal is: `bam` missed the `data_lines` and `comment_lines` properties of `sam`.

File: lib/galaxy/datatypes/binary.py [hunk 1]
```diff
@@ -331,22 +331,67 @@ class GenericAsn1Binary(Binary):
     edam_data = "data_0849"
 
 
-class BamNative(CompressedArchive):
+class _BamOrSam():
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]:
```suggestion
class _BamOrSam:
```
?

Comment (bernt-matthias) [reply lib/galaxy/datatypes/binary.py]: Odd that the linter did not complain. 

One thing that I was also wondering is why I can't specify the `MetadataElement`s in this class, but need to duplicate them in the sam and bam classes?

**Pull request merged**